File: 329.矩阵中的最长递增路径.cpp
/*
 * @lc app=leetcode.cn id=329 lang=cpp
 *
 * [329] 矩阵中的最长递增路径
 */

// @lc code=start
#include<vector>
using namespace std;

class Solution {
public:
    vector<vector<pair<int, int>>> _adjlist;
    vector<int> _cache;
    int m, n;

    vector<pair<int, int>>& adjlist(const pair<int, int>& x){
        return _adjlist[x.first * n + x.second];
    }

    int dfs(const pair<int, int>& x) {
        /////////////////////////////////////////
        if(_cache[x.first * n + x.second]) return _cache[x.first * n + x.second];
        /////////////////////////////////////////
        int ret = 0;
        for(auto& i: adjlist(x)){
            ret = max(ret, dfs({i.first, i.second}));
        }
        /////////////////////////////////////////
        _cache[x.first * n + x.second] = ret + 1;
        /////////////////////////////////////////
        return ret + 1;
    }

    int longestIncreasingPath(vector<vector<int>>& matrix) {
        m = matrix.size(), n = matrix.front().size();
        _adjlist.resize(m * n, {});
        _cache.resize(m * n, 0);

        for(int i=1; i<m; ++i){
            for(int j=0; j<n; ++j){
                if(matrix[i-1][j] > matrix[i][j]) adjlist({i, j}).push_back({i-1, j});
                else if(matrix[i-1][j] < matrix[i][j]) adjlist({i-1, j}).push_back({i, j});
            }
        }
        for(int i=0; i<m; ++i){
            for(int j=1; j<n; ++j){
                if(matrix[i][j-1] > matrix[i][j]) adjlist({i, j}).push_back({i, j-1});
                else if(matrix[i][j-1] < matrix[i][j]) adjlist({i, j-1}).push_back({i, j});
            }
        }

        int ret = 0;
        for(int i=0; i<m; ++i){
            for(int j=0; j<n; ++j){
                ret = max(ret, dfs({i, j}));
            }
        }
        return ret;
    }
};
// @lc code=end

File: 329.矩阵中的最长递增路径.py
#
# @lc app=leetcode.cn id=329 lang=python3
#
# [329] 矩阵中的最长递增路径
#

# @lc code=start
from typing import List, Tuple
from collections import defaultdict
from functools import lru_cache

class Solution:
    @lru_cache(None)
    def dfs(self, start: Tuple[int, int]) -> int:
        if start not in self.adjlist: return 1
        return max(
            self.dfs(neibour)
            for neibour in self.adjlist[start[0], start[1]]
        ) + 1

    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        m, n = len(matrix), len(matrix[0])
        self.adjlist = adjlist = defaultdict(list)
        for i in range(1, m):
            for j in range(n):
                if matrix[i-1][j] > matrix[i][j]: adjlist[i, j].append((i-1, j))
                elif matrix[i-1][j] < matrix[i][j]: adjlist[i-1, j].append((i, j))
        for i in range(m):
            for j in range(1, n):
                if matrix[i][j-1] > matrix[i][j]: adjlist[i, j].append((i, j-1))
                elif matrix[i][j-1] < matrix[i][j]: adjlist[i, j-1].append((i, j))
        return max(
            (self.dfs(candidates)
            for candidates in adjlist),
            default=1
        )

# @lc code=end

print(Solution().longestIncreasingPath(matrix = [[9,9,4],[6,6,8],[2,1,1]]))
print(Solution().longestIncreasingPath(matrix = [[3,4,5],[3,2,6],[2,2,1]]))
print(Solution().longestIncreasingPath(matrix = [[1]]))